Python/specialize.c

```diff
@@ -2560,14 +2560,36 @@ static int
 binary_op_extended_specialization(PyObject *lhs, PyObject *rhs, int oparg,
                                   _PyBinaryOpSpecializationDescr **descr)
 {
-    size_t n = sizeof(binaryop_extend_descrs)/sizeof(_PyBinaryOpSpecializationDescr);
-    for (size_t i = 0; i < n; i++) {
-        _PyBinaryOpSpecializationDescr *d = &binaryop_extend_descrs[i];
+    typedef _PyBinaryOpSpecializationDescr descr_type;
+    size_t size = Py_ARRAY_LENGTH(binaryop_extend_builtins);
+    for (size_t i = 0; i < size; i++) {
+        descr_type *d = (descr_type *)&binaryop_extend_builtins[i];
+        assert(d != NULL);
+        assert(d->guard != NULL);
         if (d->oparg == oparg && d->guard(lhs, rhs)) {
             *descr = d;
             return 1;
         }
     }
+
+    PyTypeObject *lhs_type = Py_TYPE(lhs);
```

Comment:
As I think we had discussed this offline, this logic is a bit different from operator overloading like `__add__`, since there the `rhs` can also come into play.  For this first "proof of concept" with `BINARY_SUBSCR`/`__getitem__`, it doesn't matter, but maybe leave a comment here for the future where we start to add more specializations?  It's not critical, since the difference (I think) would just mean a lost potential optimization, not a different in behaviour.